OpenTrons/opentrons-api
# fix(protocol-engine): Support door open/close during error recovery

# Overview

This corrects the behavior when you open the door in the middle of error recovery. Closes EXEC-383.

# Test plan

1. Enter error recovery.
2. Open the door. The error recovery UI should still be active, but there should be a modal saying to close the door.
3. Close the door. The error recovery UI should give you some kind of "confirm door closed and continue" button. We have not implemented this on the frontend yet (EXEC-565). You can manually issue an HTTP `play` action instead.
4. Error recovery should continue as normal after you click the "confirm and continue" button (once EXEC-565 is implemented) or issue an HTTP `play` action.

# Details

This introduces two new run `status`es to accompany the existing `awaiting-recovery` status:

1. `awaiting-recovery` (already existed)
2. `awaiting-recovery-blocked-by-open-door` (new)
3. `awaiting-recovery-paused` (new)

```mermaid
stateDiagram-v2
    awaiting_recovery: awaiting-recovery
    awaiting_recovery_paused: awaiting-recovery-paused
    awaiting_recovery_blocked_by_open_door: awaiting-recovery-blocked-by-open-door
    stop_requested: stop-requested etc.
    running: running etc.
    
    [*] --> awaiting_recovery: Command fails recoverably
    awaiting_recovery --> awaiting_recovery: HTTP `play` action
    awaiting_recovery --> awaiting_recovery_blocked_by_open_door: Door opens
    awaiting_recovery --> stop_requested: HTTP `stop` action
    awaiting_recovery_blocked_by_open_door --> awaiting_recovery_paused: Door closes
    awaiting_recovery_blocked_by_open_door --> stop_requested: HTTP `stop` action
    awaiting_recovery_paused --> awaiting_recovery: HTTP `play` action
    awaiting_recovery_paused --> stop_requested: HTTP `stop` action

    awaiting_recovery --> running: HTTP `resume-from-recovery` action
    running --> [*]
    stop_requested --> [*]
```

Having all 3 is necessary to implement the same behavior that we have for the main protocol part of the run, where after closing the door you need to click a resume button to get the robot moving again.

We can't reuse the existing `blocked-by-open-door` and `paused` statuses because various parts of the app look at the status to determine whether to be in recovery mode. We need to maintain an unambiguous `awaiting-recovery-*` status until error recovery is complete.

# Review requests

* Is adding these new statuses a good idea, or do we actually want to bite the bullet and split this monolithic field up into multiple smaller ones now?
* Do my frontend changes seem correct? Are there any other places that I missed?
* Do my docstrings on `EngineStatus` look correct?
* I suspect there's a good opportunity to test this more exhaustively with Hypothesis. Does anyone have specific ideas for what invariants we would use?

# Risk assessment

Medium. There are a lot of things on the frontend that look at `status`, and we have to carefully update them all. Consequences for getting it wrong range from buttons being enabled/disabled at the wrong times, to the screen showing an entirely wrong flow.
